File: gauntlet/go/jsontree.go
package gauntlet

// Hand-written JSON tree builders — the bar for examples/json/tree.oro.
//
// THREE FORMS, and the first two are the whole question:
//
//	TreeRec    recursive descent into a linked node struct — what a person
//	           writes, and what ADR 0014 says we may not
//	TreeFlat   a flat node table plus indices with an explicit stack — our
//	           shape, hand-written, with no clamping
//	GenMeasure emitted from the .oro
//
// TreeRec against TreeFlat prices the REPRESENTATION: pointers and per-node
// allocation against one flat array and an explicit stack. TreeFlat against
// GenMeasure prices our CODE GENERATION plus the clamped addressing the
// refinement layer asks for, with the representation held fixed.
//
// All three produce the same tree and the same answer: seen*1000 + Σ tag×depth
// over a depth-first walk. Tags are 1 number, 2 string, 3 literal, 4 array,
// 5 object; an object's children are its keys and values in document order,
// because `:` and `,` are skipped rather than structured — which is what the
// .oro does and what these must therefore do too.

const (
	treeNMax = 512
	treeDMax = 32
)

// ---------------------------------------------------------- recursive, boxed

type jnode struct {
	tag, val int
	kid, sib *jnode
}

func TreeRec(src []int) int {
	root, _ := jparseValue(src, jskip(src, 0))
	seen, acc := jwalkRec(root, 1)
	return seen*1000 + acc
}

func jskip(src []int, i int) int {
	for i < len(src) && (src[i] == 32 || src[i] == 9 || src[i] == 10 || src[i] == 13 ||
		src[i] == 58 || src[i] == 44) {
		i++
	}
	return i
}

func jparseValue(src []int, i int) (*jnode, int) {
	if i >= len(src) {
		return nil, i
	}
	c := src[i]
	if c == 123 || c == 91 {
		n := &jnode{tag: 5}
		if c == 91 {
			n.tag = 4
		}
		i++
		var last *jnode
		for {
			i = jskip(src, i)
			if i >= len(src) || src[i] == 125 || src[i] == 93 {
				if i < len(src) {
					i++
				}
				break
			}
			child, ni := jparseValue(src, i)
			if child == nil {
				break
			}
			i = ni
			if last == nil {
				n.kid = child
			} else {
				last.sib = child
			}
			last = child
		}
		return n, i
	}
	if c == 34 {
		ni := tokStringI(src, i)
		return &jnode{tag: 2, val: ni - i}, ni
	}
	if tokNumeric(c) {
		j := i
		for j < len(src) && tokNumeric(src[j]) {
			j++
		}
		return &jnode{tag: 1, val: j - i}, j
	}
	if tokAlpha(c) {
		j := i
		for j < len(src) && tokAlpha(src[j]) {
			j++
		}
		return &jnode{tag: 3, val: j - i}, j
	}
	return nil, i + 1
}

func jwalkRec(n *jnode, d int) (int, int) {
	if n == nil {
		return 0, 0
	}
	seen, acc := 1, n.tag*d
	for c := n.kid; c != nil; c = c.sib {
		s, a := jwalkRec(c, d+1)
		seen += s
		acc += a
	}
	return seen, acc
}

// ------------------------------------------------------------- flat, indexed
//
// The same algorithm the .oro runs, written the way a person would write it:
// no clamped addressing, because Go's own bounds check is already there and a
// person would rely on it.

func TreeFlat(src []int) int {
	nodes := make([]int, 4*treeNMax)
	stk := make([]int, 2*treeDMax)
	i, nn, sp := 0, 1, 0
	for {
		if i >= len(src) || nn >= treeNMax || sp >= treeDMax {
			break
		}
		c := src[i]
		switch {
		case c == 32 || c == 9 || c == 10 || c == 13 || c == 58 || c == 44:
			i++
		case c == 123 || c == 91:
			tg := 5
			if c == 91 {
				tg = 4
			}
			nodes[4*nn] = tg
			nodes[4*nn+1] = 0
			jlink(nodes, stk, sp, nn)
			if sp >= 1 {
				stk[2*(sp-1)+1] = nn
			}
			stk[2*sp] = nn
			stk[2*sp+1] = 0
			i++
			sp++
			nn++
		case c == 125 || c == 93:
			i++
			if sp >= 1 {
				sp--
			}
		case c == 34 || tokNumeric(c) || tokAlpha(c):
			tg, ni := 2, 0
			switch {
			case c == 34:
				ni = tokStringI(src, i)
			case tokNumeric(c):
				tg = 1
				j := i
				for j < len(src) && tokNumeric(src[j]) {
					j++
				}
				ni = j
			default:
				tg = 3
				j := i
				for j < len(src) && tokAlpha(src[j]) {
					j++
				}
				ni = j
			}
			nodes[4*nn] = tg
			nodes[4*nn+1] = ni - i
			jlink(nodes, stk, sp, nn)
			if sp >= 1 {
				stk[2*(sp-1)+1] = nn
			}
			i = ni
			nn++
		default:
			i++
		}
	}
	return jwalkFlat(nodes)
}

func jlink(nodes, stk []int, sp, k int) {
	if sp < 1 {
		return
	}
	if lc := stk[2*(sp-1)+1]; lc == 0 {
		nodes[4*stk[2*(sp-1)]+2] = k
	} else {
		nodes[4*lc+3] = k
	}
}

func jwalkFlat(nodes []int) int {
	wl := make([]int, 2*treeNMax)
	wl[0], wl[1] = 1, 1
	sp, seen, acc, steps := 1, 0, 0, 0
	for sp >= 1 && steps < 2*treeNMax {
		n, d := wl[2*(sp-1)], wl[2*(sp-1)+1]
		sb, kd := nodes[4*n+3], nodes[4*n+2]
		sp--
		if sb != 0 {
			wl[2*sp], wl[2*sp+1] = sb, d
			sp++
		}
		if kd != 0 {
			wl[2*sp], wl[2*sp+1] = kd, d+1
			sp++
		}
		seen++
		acc += nodes[4*n] * d
		steps++
	}
	return seen*1000 + acc
}

// THE SAME PROGRAM WITH THE CLAMPS, which is what the .oro carries and the
// hand-written version above does not.
//
// The refinement layer discharges every index obligation by CLAMPING rather
// than by proving: `(go.* 4 (cn k))` is in range by construction, at the price
// of two compares per access. A person writing this by hand relies on Go's own
// bounds check instead. If this measures like the generated code then the
// clamps are the whole gap; if it measures like TreeFlat they are not.
//
// json-tokeniser found the same layer's three extra compares cost nothing —
// but that program had three of them and this one has one per table access.

func jcn(k int) int {
	if k < 0 || k >= treeNMax {
		return 0
	}
	return k
}

func jcd(d int) int {
	if d < 0 || d >= treeDMax {
		return 0
	}
	return d
}

func jcw(k int) int {
	if k < 0 || k >= treeNMax {
		return 0
	}
	return k
}

func TreeFlatClamped(src []int) int {
	nodes := make([]int, 4*treeNMax)
	stk := make([]int, 2*treeDMax)
	i, nn, sp := 0, 1, 0
	for {
		if i >= len(src) || nn >= treeNMax || sp >= treeDMax {
			break
		}
		c := src[i]
		switch {
		case c == 32 || c == 9 || c == 10 || c == 13 || c == 58 || c == 44:
			i++
		case c == 123 || c == 91:
			tg := 5
			if c == 91 {
				tg = 4
			}
			nodes[4*jcn(nn)] = tg
			nodes[4*jcn(nn)+1] = 0
			jlinkClamped(nodes, stk, sp, nn)
			if sp >= 1 {
				stk[2*jcd(sp-1)+1] = nn
			}
			stk[2*jcd(sp)] = nn
			stk[2*jcd(sp)+1] = 0
			i++
			sp++
			nn++
		case c == 125 || c == 93:
			i++
			if sp >= 1 {
				sp--
			}
		case c == 34 || tokNumeric(c) || tokAlpha(c):
			tg, ni := 2, 0
			switch {
			case c == 34:
				ni = tokStringI(src, i)
			case tokNumeric(c):
				tg = 1
				j := i
				for j < len(src) && tokNumeric(src[j]) {
					j++
				}
				ni = j
			default:
				tg = 3
				j := i
				for j < len(src) && tokAlpha(src[j]) {
					j++
				}
				ni = j
			}
			nodes[4*jcn(nn)] = tg
			nodes[4*jcn(nn)+1] = ni - i
			jlinkClamped(nodes, stk, sp, nn)
			if sp >= 1 {
				stk[2*jcd(sp-1)+1] = nn
			}
			i = ni
			nn++
		default:
			i++
		}
	}
	return jwalkFlatClamped(nodes)
}

func jlinkClamped(nodes, stk []int, sp, k int) {
	if sp < 1 {
		return
	}
	if lc := stk[2*jcd(sp-1)+1]; lc == 0 {
		nodes[4*jcn(stk[2*jcd(sp-1)])+2] = k
	} else {
		nodes[4*jcn(lc)+3] = k
	}
}

func jwalkFlatClamped(nodes []int) int {
	wl := make([]int, 2*treeNMax)
	wl[0], wl[1] = 1, 1
	sp, seen, acc, steps := 1, 0, 0, 0
	for sp >= 1 && steps < 2*treeNMax {
		n, d := wl[2*jcw(sp-1)], wl[2*jcw(sp-1)+1]
		sb, kd := nodes[4*jcn(n)+3], nodes[4*jcn(n)+2]
		sp--
		if sb != 0 {
			wl[2*jcw(sp)], wl[2*jcw(sp)+1] = sb, d
			sp++
		}
		if kd != 0 {
			wl[2*jcw(sp)], wl[2*jcw(sp)+1] = kd, d+1
			sp++
		}
		seen++
		acc += nodes[4*jcn(n)] * d
		steps++
	}
	return seen*1000 + acc
}

File: gauntlet/js/jsontree.mjs
// A JSON tree on JavaScript: emitted against hand-written.
//
// THREE FORMS, and the first two are the question:
//
//   rec    recursive descent into linked objects — what a person writes
//   flat   a flat node table plus indices with an explicit stack — our shape
//   NATIVE emitted from the .oro
//
// `rec` against `flat` prices the REPRESENTATION on this host, and it is not
// the same question as on Go. V8 allocates in a bump-pointer nursery and
// collects short-lived objects almost for free, so the 2.52x the flat form won
// on Go is a measurement rather than a principle (ADR 0008), and it has to be
// taken again here.
//
// `flat` against NATIVE prices our code generation with the representation held
// fixed.
//
//   node jsontree.mjs --check     correctness only
//   node jsontree.mjs <case>      run one case
//
// ONE PROCESS PER COMPARISON, and NAMED imports on both sides — V8 carries
// optimization state across benchmarks in a process, and a namespace property
// load costs 1.66x (native-js-2026-08-20).

import { nativeMeasure } from "./gen_jsontree_native.mjs";

const NMAX = 512;
const DMAX = 32;

const isNum = (c) =>
  (c >= 48 && c <= 57) || c === 45 || c === 43 || c === 46 || c === 101 || c === 69;
const isAlpha = (c) => c >= 97 && c <= 122;
const isSkip = (c) =>
  c === 32 || c === 9 || c === 10 || c === 13 || c === 58 || c === 44;

function scanString(a, i) {
  let j = i + 1;
  for (;;) {
    if (j >= a.length) return j;
    if (a[j] === 92) { j += 2; continue; }
    if (a[j] === 34) return j + 1;
    j++;
  }
}

// ------------------------------------------------------------------- boxed

function skip(a, i) {
  while (i < a.length && isSkip(a[i])) i++;
  return i;
}

function parseValue(a, i) {
  if (i >= a.length) return [null, i];
  const c = a[i];
  if (c === 123 || c === 91) {
    const n = { tag: c === 123 ? 5 : 4, val: 0, kid: null, sib: null };
    i++;
    let last = null;
    for (;;) {
      i = skip(a, i);
      if (i >= a.length || a[i] === 125 || a[i] === 93) {
        if (i < a.length) i++;
        break;
      }
      const [child, ni] = parseValue(a, i);
      if (child === null) break;
      i = ni;
      if (last === null) n.kid = child; else last.sib = child;
      last = child;
    }
    return [n, i];
  }
  if (c === 34) { const ni = scanString(a, i); return [{ tag: 2, val: ni - i, kid: null, sib: null }, ni]; }
  if (isNum(c)) {
    let j = i;
    while (j < a.length && isNum(a[j])) j++;
    return [{ tag: 1, val: j - i, kid: null, sib: null }, j];
  }
  if (isAlpha(c)) {
    let j = i;
    while (j < a.length && isAlpha(a[j])) j++;
    return [{ tag: 3, val: j - i, kid: null, sib: null }, j];
  }
  return [null, i + 1];
}

function walkRec(n, d) {
  if (n === null) return [0, 0];
  let seen = 1, acc = n.tag * d;
  for (let c = n.kid; c !== null; c = c.sib) {
    const [s, a] = walkRec(c, d + 1);
    seen += s; acc += a;
  }
  return [seen, acc];
}

export function treeRec(a) {
  const [root] = parseValue(a, skip(a, 0));
  const [seen, acc] = walkRec(root, 1);
  return seen * 1000 + acc;
}

// -------------------------------------------------------------------- flat

export function treeFlat(a) {
  const nodes = new Array(4 * NMAX).fill(0);
  const stk = new Array(2 * DMAX).fill(0);
  let i = 0, nn = 1, sp = 0;
  for (;;) {
    if (i >= a.length || nn >= NMAX || sp >= DMAX) break;
    const c = a[i];
    if (isSkip(c)) { i++; continue; }
    if (c === 123 || c === 91) {
      nodes[4 * nn] = c === 123 ? 5 : 4;
      nodes[4 * nn + 1] = 0;
      link(nodes, stk, sp, nn);
      if (sp >= 1) stk[2 * (sp - 1) + 1] = nn;
      stk[2 * sp] = nn;
      stk[2 * sp + 1] = 0;
      i++; sp++; nn++;
      continue;
    }
    if (c === 125 || c === 93) { i++; if (sp >= 1) sp--; continue; }
    if (c === 34 || isNum(c) || isAlpha(c)) {
      let tg = 2, ni = 0;
      if (c === 34) { ni = scanString(a, i); }
      else if (isNum(c)) { tg = 1; let j = i; while (j < a.length && isNum(a[j])) j++; ni = j; }
      else { tg = 3; let j = i; while (j < a.length && isAlpha(a[j])) j++; ni = j; }
      nodes[4 * nn] = tg;
      nodes[4 * nn + 1] = ni - i;
      link(nodes, stk, sp, nn);
      if (sp >= 1) stk[2 * (sp - 1) + 1] = nn;
      i = ni; nn++;
      continue;
    }
    i++;
  }
  return walkFlat(nodes);
}

function link(nodes, stk, sp, k) {
  if (sp < 1) return;
  const lc = stk[2 * (sp - 1) + 1];
  if (lc === 0) nodes[4 * stk[2 * (sp - 1)] + 2] = k;
  else nodes[4 * lc + 3] = k;
}

function walkFlat(nodes) {
  const wl = new Array(2 * NMAX).fill(0);
  wl[0] = 1; wl[1] = 1;
  let sp = 1, seen = 0, acc = 0, steps = 0;
  while (sp >= 1 && steps < 2 * NMAX) {
    const n = wl[2 * (sp - 1)], d = wl[2 * (sp - 1) + 1];
    const sb = nodes[4 * n + 3], kd = nodes[4 * n + 2];
    sp--;
    if (sb !== 0) { wl[2 * sp] = sb; wl[2 * sp + 1] = d; sp++; }
    if (kd !== 0) { wl[2 * sp] = kd; wl[2 * sp + 1] = d + 1; sp++; }
    seen++; acc += nodes[4 * n] * d; steps++;
  }
  return seen * 1000 + acc;
}

// ------------------------------------------------------------------- input

function makeDoc(records) {
  let s = '{"items":[';
  for (let r = 0; r < records; r++) {
    if (r > 0) s += ",";
    s += '{"id":1234,"name":"a b\\"c","tags":["x","y","z"],' +
      '"score":-12.5e3,"ok":true,"prev":null,' +
      '"meta":{"depth":2,"flag":false}}';
  }
  return s + "]}";
}

const bytesOf = (t) => Array.from(t, (ch) => ch.charCodeAt(0));
const doc = bytesOf(makeDoc(20));

function check() {
  let bad = 0;
  for (const n of [0, 1, 2, 5, 20]) {
    const a = bytesOf(makeDoc(n));
    const want = treeRec(a);
    for (const [nm, got] of [["flat", treeFlat(a)], ["NATIVE", nativeMeasure(a)]]) {
      if (got !== want) { console.log(`records=${n} ${nm}: ${got} != ${want}`); bad++; }
    }
  }
  for (const s of ["[1,2]", '{"a":1}', "[[1],2]", '{"a":[1,2],"b":true}',
                   "[]", "{}", "[[[[1]]]]", '["a\\"b"]']) {
    const a = bytesOf(s);
    const want = treeRec(a);
    for (const [nm, got] of [["flat", treeFlat(a)], ["NATIVE", nativeMeasure(a)]]) {
      if (got !== want) { console.log(`${JSON.stringify(s)} ${nm}: ${got} != ${want}`); bad++; }
    }
  }
  console.log(bad === 0 ? "all three agree" : `${bad} disagreement(s)`);
  if (bad) process.exit(1);
}

function bench(name, fn, iters, runs = 7) {
  const w0 = process.hrtime.bigint();
  for (let i = 0; i < 20000; i++) {
    fn();
    if (i >= 1000 && Number(process.hrtime.bigint() - w0) > 2e9) break;
  }
  const ts = [];
  for (let r = 0; r < runs; r++) {
    const t0 = process.hrtime.bigint();
    for (let i = 0; i < iters; i++) fn();
    ts.push(Number(process.hrtime.bigint() - t0) / iters);
  }
  ts.sort((a, b) => a - b);
  console.log(`${name.padEnd(32)} ${ts[ts.length >> 1].toFixed(1).padStart(12)} ns/op`);
}

const CASES = {
  rec: ["T  tree recursive     hand", () => treeRec(doc), 2000],
  flat: ["T  tree flat          hand", () => treeFlat(doc), 2000],
  native: ["T  tree flat          NATIVE", () => nativeMeasure(doc), 2000],
};

const which = process.argv[2];
if (which === "--check" || which === undefined) check();
else {
  const c = CASES[which];
  if (!c) { console.error(`no such case: ${which}`); process.exit(1); }
  bench(c[0], c[1], c[2]);
}
